### ISSUE TITLE
sql: panic due to memory account incorrectly left open after disconnecting all nodes during export

### ISSUE THREAD

**Issue by a-robinson** [cockroachdb/cockroach#26112]
I started a local 3 node cluster using roachprod (`roachprod start local`), imported the startrek database to them (`cockroach gen example-data startrek | cockroach sql`), set up an enterprise license and gcs access credentials, then ran an EXPORT [1] and shut off my computer's wifi (which not only disconnected the nodes from GCS, but from each other because they were all using a hostname defined on the wifi interface). After about 14 seconds, [2] happened.

[1] 
```
EXPORT INTO CSV 'gs://tmp829410720848561/csv6?AUTH=default' FROM SELECT * from startrek.quotes as q1 cross join startrek.quotes as q2 ORDER BY q1.stardate DESC;
```

[2]
```
I180525 22:56:15.221316 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] distsql: bytes usage increases to 1.0 MiB (+40960)
I180525 22:56:15.221355 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] flow: bytes usage increases to 1.0 MiB (+40960)
I180525 22:56:15.221647 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] sortall-limited: bytes usage increases to 1.0 MiB (+40960)
I180525 22:56:15.223088 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] distsql: bytes usage increases to 2.0 MiB (+10240)
I180525 22:56:15.223096 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] flow: bytes usage increases to 2.0 MiB (+10240)
I180525 22:56:15.223267 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] sortall-limited: bytes usage increases to 2.0 MiB (+71680)
I180525 22:56:15.226093 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] distsql: bytes usage increases to 4.0 MiB (+10240)
I180525 22:56:15.226099 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] flow: bytes usage increases to 4.0 MiB (+10240)
I180525 22:56:15.226339 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] sortall-limited: bytes usage increases to 4.0 MiB (+10240)
I180525 22:56:15.232810 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] distsql: bytes usage increases to 8.0 MiB (+10240)
I180525 22:56:15.232819 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] flow: bytes usage increases to 8.0 MiB (+10240)
I180525 22:56:15.233178 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] sortall-limited: bytes usage increases to 8.0 MiB (+10240)
I180525 22:56:15.248098 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] distsql: bytes usage increases to 16 MiB (+10240)
I180525 22:56:15.248108 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] flow: bytes usage increases to 16 MiB (+10240)
I180525 22:56:15.248428 1353 util/mon/bytes_usage.go:576  [client=[::1]:57310,user=root,n1,sortAll/3] sortall-limited: bytes usage increases to 16 MiB (+10240)
I180525 22:56:19.836109 108 server/status/runtime.go:219  [n1] runtime stats: 151 MiB RSS, 210 goroutines, 147 MiB/4.0 MiB/171 MiB GO alloc/idle/total, 21 MiB/30 MiB CGO alloc/total, 56.38cgo/sec, 0.03/0.
01 %(u/s)time, 0.00 %gc (1x)
I180525 22:56:24.839218 219 storage/node_liveness.go:773  [n1,hb] retrying liveness update after storage.errRetryLiveness: result is ambiguous (context finished during distsender send: context deadline ex
ceeded)
W180525 22:56:24.839291 219 storage/node_liveness.go:528  [n1,hb] slow heartbeat took 4.5s
W180525 22:56:24.839321 219 storage/node_liveness.go:465  [n1,hb] failed node liveness heartbeat: context deadline exceeded
W180525 22:56:24.979567 121 storage/raft_transport.go:465  [n1] raft transport stream to node 3 failed: rpc error: code = Unavailable desc = transport is closing
W180525 22:56:24.979595 249 vendor/google.golang.org/grpc/clientconn.go:1293  grpc: addrConn.createTransport failed to connect to {alex-laptop.local:26261 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
W180525 22:56:24.979633 249 vendor/google.golang.org/grpc/clientconn.go:1429  grpc: addrConn.transportMonitor exits due to: grpc: the connection is closing
I180525 22:56:25.734328 1413 storage/raft_transport.go:459  [n1] raft transport stream to node 3 established
W180525 22:56:25.945256 278 vendor/google.golang.org/grpc/clientconn.go:1293  grpc: addrConn.createTransport failed to connect to {alex-laptop.local:26259 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
W180525 22:56:25.945363 278 vendor/google.golang.org/grpc/clientconn.go:1429  grpc: addrConn.transportMonitor exits due to: context canceled
E180525 22:56:25.945403 1401 storage/replica_range_lease.go:292  [n1,s1,r4/1:/System/{NodeLive…-tsd}] not incrementing epoch on n2 because next leaseholder (n1) not live (err = <nil>)
I180525 22:56:26.831858 1461 storage/raft_transport.go:459  [n1] raft transport stream to node 2 established
W180525 22:56:26.963653 219 storage/node_liveness.go:528  [n1,hb] slow heartbeat took 2.1s
W180525 22:56:27.154825 633 vendor/google.golang.org/grpc/clientconn.go:1293  grpc: addrConn.createTransport failed to connect to {alex-laptop.local:26257 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
W180525 22:56:27.154844 633 vendor/google.golang.org/grpc/clientconn.go:1429  grpc: addrConn.transportMonitor exits due to: grpc: the connection is closing
I180525 22:56:29.831867 108 server/status/runtime.go:219  [n1] runtime stats: 155 MiB RSS, 196 goroutines, 98 MiB/52 MiB/171 MiB GO alloc/idle/total, 21 MiB/30 MiB CGO alloc/total, 72.13cgo/sec, 0.01/0.01 %(u/s)time, 0.00 %gc (1x)
W180525 22:56:33.657931 110 server/node.go:802  [n1,summaries] health alerts detected: {Alerts:[{StoreID:0 Category:METRICS Description:liveness.heartbeatfailures Value:1}]}
panic: txn: unexpected 81920 leftover bytes [recovered]
        panic: txn: unexpected 81920 leftover bytes

goroutine 787 [running]:
github.com/cockroachdb/cockroach/pkg/util/log.ReportOrPanic(0x6436200, 0xc425aec880, 0xc4206fc000, 0xc424edc2d0, 0x24, 0x0, 0x0, 0x0)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:494 +0x211
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc4260b6320, 0x6436200, 0xc425aec880, 0xc427412b01)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:398 +0x263
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(0xc4260b6320, 0x6436200, 0xc425aec880)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:384 +0x44
github.com/cockroachdb/cockroach/pkg/sql.(*txnState).finishSQLTxn(0xc425dac088, 0x6436200, 0xc420a9ccc0)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/txn_state.go:233 +0x59
github.com/cockroachdb/cockroach/pkg/sql.cleanupAndFinish(0x6436200, 0xc420a9ccc0, 0x6415660, 0xc420317980, 0x5f53580, 0xc425dac088, 0x6415500, 0xc42028f200, 0x5e73c20, 0xc42028f210, ...)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_fsm.go:513 +0xd8
github.com/cockroachdb/cockroach/pkg/util/fsm.Transitions.apply(0xc4200a0a20, 0x6436200, 0xc420a9ccc0, 0x6415660, 0xc420317980, 0x5f53580, 0xc425dac088, 0x6415500, 0xc42028f200, 0x5e73c20, ...)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/fsm/fsm.go:107 +0x18c
github.com/cockroachdb/cockroach/pkg/util/fsm.(*Machine).ApplyWithPayload(0xc425dac060, 0x6436200, 0xc420a9ccc0, 0x6415500, 0xc42028f200, 0x5e73c20, 0xc42028f210, 0x0, 0x60cc8ea)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/fsm/fsm.go:136 +0x10c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).close(0xc425dac000, 0x6436200, 0xc420a9ccc0, 0xc427412f00)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:680 +0x1cf
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc425dac000, 0x6436200, 0xc420a9ccc0, 0x5d0ce60, 0xc42028f110)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:642 +0x226
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc425dac000, 0x6436200, 0xc420a9ccc0)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:399 +0x61
panic(0x5d0ce60, 0xc42028f110)
        /usr/local/go/src/runtime/panic.go:505 +0x229
github.com/cockroachdb/cockroach/pkg/util/log.ReportOrPanic(0x6436200, 0xc425aec880, 0xc4206fc000, 0xc424edc270, 0x24, 0x0, 0x0, 0x0)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:494 +0x211
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc4260b6320, 0x6436200, 0xc425aec880, 0x1)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:398 +0x263
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(0xc4260b6320, 0x6436200, 0xc425aec880)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:384 +0x44
github.com/cockroachdb/cockroach/pkg/sql.(*txnState).finishSQLTxn(0xc425dac088, 0x64362c0, 0xc4253506f0)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/txn_state.go:233 +0x59
github.com/cockroachdb/cockroach/pkg/sql.cleanupAndFinish(0x64362c0, 0xc4253506f0, 0x6415660, 0xc420317980, 0x5f53580, 0xc425dac088, 0x6415500, 0xc42028ebe0, 0x5e73c20, 0xc42028ebf0, ...)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_fsm.go:513 +0xd8
github.com/cockroachdb/cockroach/pkg/util/fsm.Transitions.apply(0xc4200a0a20, 0x64362c0, 0xc4253506f0, 0x6415660, 0xc420317980, 0x5f53580, 0xc425dac088, 0x6415500, 0xc42028ebe0, 0x5e73c20, ...)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/fsm/fsm.go:107 +0x18c
github.com/cockroachdb/cockroach/pkg/util/fsm.(*Machine).ApplyWithPayload(0xc425dac060, 0x64362c0, 0xc4253506f0, 0x6415500, 0xc42028ebe0, 0x5e73c20, 0xc42028ebf0, 0x6415500, 0xc42028ebe0)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/fsm/fsm.go:136 +0x10c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).txnStateTransitionsApplyWrapper(0xc425dac000, 0x6415500, 0xc42028ebe0, 0x5e73c20, 0xc42028ebf0, 0x89bc148, 0xc426f36000, 0x43, 0x0, 0x0, ...)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1808 +0x13b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc425dac000, 0x6436200, 0xc420a9ccc0, 0xc4203e04f0, 0x0, 0x0)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1177 +0xc8b
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc420bb0000, 0x6436200, 0xc420a9ccc0, 0x0, 0x0, 0xc425f3d050, 0x4, 0xc425f3d03d, 0xd, 0x641d940, ...)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:401 +0x1bb
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func3(0xc420bb0000, 0x6436200, 0xc420a9ccc0, 0xc4255e0000, 0x5400, 0x15000, 0xc420a16600, 0xc4203e04f0, 0xc4203e04e0, 0xc42040a7e0, ...)
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:266 +0x122
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
        /Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:265 +0xf27
```

Found as part of #25223

**Comment by a-robinson** [comment]
For what it's worth, the same thing appears to happen even if the nodes don't lose connectivity with each other.

**Comment by knz** [comment]
This is a case of an error path (upon node disconnect) not properly closing memory monitor accounts.

**Comment by mjibson** [comment]
I thought this would have been fixed by #26029 but it doesn't appear to have been. Blergh.

**Comment by mjibson** [comment]
I can no longer reproduce this. I'm going to close it unless someone else is able to.
